### PR TITLE
Always select non link-local network interface, fixes #11961.

### LINKED ISSUE #11961: Kubernetes endpoint should not use link-local address of interface
On certain cloud providers like digital ocean the public interface has multiple addresses, one of which is a link-local address that points to a metadata server. The node is supposed to come up and contact the metadata server to bootstrap the public interface with a real address. At this point, the interface has 2 addresses, eg (from irc debugging): https://dpaste.de/T8s0/raw. 

Looks like we create the kubernets endpoint using the first, link-local address and it always gets rejected. The kubernetes master service creation can keep failing (and re-checking for a valid public ip) when this happens because it knows link-local is not allowed: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/master/controller.go#L97

### PR DESCRIPTION
**Pull request by paralin**
Fixes #11961. The API server will refuse to start if a link-local IP address is selected, however, ChooseNetworkInterface can still select link-local network routes. This causes a startup failure on platforms like DigitalOcean that use a link-local route to fetch metadata.

### CODE REVIEW

**Comment by paralin** [comment]
I signed it!

That bot is so fast it posted 2 replies!

**Comment by thockin** [comment]
ok to test

**Comment by thockin** [comment]
Code looks good.  2 remarks:

1: something funky in your 2nd commit - can you please squash to a single commit and see that it produces the minimal change
2: please add a test case in `TestChooseHostInterfaceFromRoute`

**Comment by paralin** [comment]
@thockin Cleaned up the commits, give me a few minutes to implement the test.

**Comment by paralin** [comment]
Okay @thockin I took a sample of /proc/net/route from a real DigitalOcean instance and wrote the test around that. It appears that it will select the correct IP address now.

**Comment by paralin** [comment]
Also the test failure on the last commit was due to a Shippable authentication problem with GitHub...

**Comment by paralin** [comment]
@thockin Fixed the case in the variable, fixed the /proc/net/route entry, and corrected the Addrs function in the test.

File: pkg/util/util_test.go [hunk 1]
```diff
@@ -499,6 +505,19 @@ func (_ validNetworkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
...
+type validNetworkInterfaceWithoutLinkLocal struct {
+}
+
+func (_ validNetworkInterfaceWithoutLinkLocal) InterfaceByName(intfName string) (*net.Interface, error) {
+	c := net.Interface{Index: 0, MTU: 0, Name: "eth0", HardwareAddr: nil, Flags: net.FlagUp}
+	return &c, nil
+}
+func (_ validNetworkInterfaceWithoutLinkLocal) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	var ifat []net.Addr
+	ifat = []net.Addr{addrStruct{val: "45.55.47.146/19"}, addrStruct{val: "169.254.162.166/16"}}
```
Comment (thockin) [inline pkg/util/util_test.go]: shouldn't the link-local come first, to prove it does not get selected?

Comment (paralin) [inline pkg/util/util_test.go]: Whoops.

**Comment by paralin** [comment]
@thockin For whatever reason if I invert the two in the test I'm getting an error that the code cannot select an IP address. Very very weird. I'll keep looking into it.

**Comment by paralin** [comment]
Okay the issue is in `getIPFromInterface` it just gets the first IP address (I think). Working on this.

**Comment by paralin** [comment]
Okay, the solution is to move the check into `getFinalIP`. Pushing the fix.

**Comment by paralin** [comment]
@thockin Fixes complete, the tests pass on my end. I amended the comment on `getFinalIP` since it will now skip IPs that are loopback (as before) OR link-local. This doesn't seem to affect anything but it's worth noting.

**Comment by paralin** [comment]
This apparently affected the DNS backend tests... Will investigate more tomorrow.

File: pkg/util/util_test.go [hunk 2]
```diff
@@ -499,6 +505,19 @@ func (_ validNetworkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
 	return ifat, nil
 }
 
+type validNetworkInterfaceWithoutLinkLocal struct {
```
Comment (thockin) [inline pkg/util/util_test.go]: one more nit:  s/Without/With ?

Comment (paralin) [inline pkg/util/util_test.go]: Maybe. It checks that it *doesn't* select link local, but the test really is about if the network interface is selected properly with a link-local in the pool...

e2e is broken though so I think that's the bigger issue.

**Comment by paralin** [comment]
Interesting, now it's working in the e2e tests. I think something else might have fixed them in the commit history. I'd say this is ready for merge now, @thockin

**Comment by paralin** [comment]
Changed the test from WithoutLinkLocal to WithLinkLocal

**Comment by thockin** [comment]
lgtm

**Pull request merged**